Fix #4880: Add NPS Survey Gating Logic

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fix #4880.

This is PR 2 of 6 planned PRs.

The changes introduced in this PR include:
### Tracking of Active Time in an Exploration
We are introducing a timing mechanism that kicks off when an exploration is started, and is stopped when an exploration is ended or the app goes into the background. This allows us to improve the precision of tracking when a learner is actively learning.

### Tracking of when a survey is shown on a profile
Record the last time when a profile was served the survey.

### Increased granularity in the time of day categorization.
We are breaking morning into `EARLY_MORNING` and `MID_MORNING`, and evening into `EVENING` and `LATE_NIGHT`

### Survey Gating
Using the 3 criteria tracked above, we decide who gets served with the survey.

### New Callback Pattern
We are introducing a set of daggerbound callbacks to improve sharing information more cleanly between domain controllers.

### Test file changes
Approximately 150 test files have changed to include `ExplorationprogressModule`, which is the application-scoped module used to provide dependencies needed to track active learning time.

### New Platform Parameters
We are itroducing two new platform parameters to be able to remotely tweak the minimimum aggregate time and the grace period required for survey gating.
 
### Proto Changes
- Added a `survey_last_shown_timestamp_ms` to the `profile.proto` which represents the epoch timestamp in milliseconds when the nps survey was previously shown on this profile.
- `topic.proto`
```
 // Top-level proto used to store the aggregate learning duration timestamps in a topic,
// on per-profile basis.
message TopicLearningTimeDatabase {
  // Map from topic ID to TopicLearningTime.
  map<string, TopicLearningTime> aggregate_topic_learning_time = 1;
}

// Represents the active learning time spent in a topic.
message TopicLearningTime {
  // The ID of the topic for which this learning time has been recorded.
  string topic_id = 1;

  // The total time spent by a learner in this topic, in milliseconds.
  int64 topic_learning_time_ms = 2;

  // Represents the last Unix epoch timestamp when the topic_aggregate_learning_time_ms was updated,
  // in milliseconds, coinciding with the last time the learner played an exploration in this topic.
  int64 last_updated_time_ms = 3;
}
```

### Domain BUILD.bazel
Added the `survey` package to the `MIGRATED_PROD_FILES` to be excluded automatically from the top-level utility library. 

### Test File Exemptions
Exempted `ExplorationProgressListener` and `ApplicationLifecycleListener` from this check.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

adhiamboperes committed:
Create survey related models

adhiamboperes committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into develop

adhiamboperes committed:
Fix protobuf formatting errors

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

adhiamboperes committed:
Add a survey_last_shown_timestamp field to the profile

adhiamboperes committed:
Add methods for saving and retrieving survey_last_shown_timestamp

adhiamboperes committed:
Add gating logic based on time of day and survey last shown date

adhiamboperes committed:
Create TopicLearningTimeDatabase

Top-level proto used to store the aggregate learning duration timestamps in a topic, per-profile basis.

adhiamboperes committed:
Add topic active exploration time logic

Created a controller that records when an exploration session has been started and paused, and also saves and retrieves the total active time spent in the exploration.

adhiamboperes committed:
Fix ktlint violations

adhiamboperes committed:
Refactor fetchSurveyLastShownTimestamp to return a default date instead of a nullable value

adhiamboperes committed:
Refactor isMoreThanRequiredDaysAgo() to improve readability.

Switched to use more descriptive naming.

adhiamboperes committed:
Add aggregate learning time gating condition

Tested that overall gating logic works as expected.

adhiamboperes committed:
Create a new dagger bound listener set that will be fired in response to an exploration started and stopped.

These will be called by controllers as a way to get event-driven information in a lifecycle-safe way.

adhiamboperes committed:
Add the new module to the dagger application components.

adhiamboperes committed:
Fix ktlint newline violations

adhiamboperes committed:
Inject and initialize ExplorationProgressListener in AbstractOppiaApplication:WIP

Used dummy default values. Work needs to be done to pass the correct defaults or change the initialization.

adhiamboperes committed:
Add implementation for the ExplorationProgressListener

Created a provider and added the module to the application components.

adhiamboperes committed:
Refactor ExplorationProgressController to fire events: WIP

ExplorationProgressController.kt will fire an ExplorationProgressListener callback whenever an exploration is started or finished.

TODO: pass the actual values to onExplorationSessionPaused()

adhiamboperes committed:
Add logic to show survey dialog when gating criteria satisfied: WIP

This logic currently lives in the presenter, but we need it to move to the controller so that the UI does not have to make this decision, but rather observe and respond to a state sent by the controller.

adhiamboperes committed:
Refactor the implementation of ExplorationProgressListeners: WIP

We need to pass the correct instances of topic ID and profile ID, and move the implementation of ApplicationLifecycleListener to it's own class.

adhiamboperes committed:
Fix: ktlint

adhiamboperes committed:
Fix formating nits

adhiamboperes committed:
Refactor MarketFitAnswer field   NA_DONT_USE_OPPIA_ANYMORE to NOT_APPLICABLE_WONT_USE_OPPIA_ANYMORE

'NA' is a bit ambiguous, and the context is "won't use Oppia anymore". If they aren't using the app, they presumably wouldn't be filling out this survey.

adhiamboperes committed:
Refactor enum constants to not include the enum name.

For proto enums, we don't need to prefix-qualify each one since they'll be part of a generated Java enum -- except the first one since the unspecified constant should always repeat the enum name per proto convention.

adhiamboperes committed:
Refactor message ID documentation

Refactored to explain the ID format and the uniqueness guaranteed.

adhiamboperes committed:
Refactor SurveyQuestionAnswer to SurveyAnswerOption

To convey that it is a multiple choice answer option.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Add ExplorationProgressModule to TestApplicationComponents

adhiamboperes committed:
Add BUILD definition for TopicLearningTimeControllerTest

adhiamboperes committed:
Add tests for TopicLearningTimeController

adhiamboperes committed:
Rename TopicLearningTimeController as ExplorationActiveTimeController

Sounds more in line with the package location and purpose.

adhiamboperes committed:
Fix ktlint

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

adhiamboperes committed:
Merge branch 'nps-survey-models' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Refactor SurveyQuestion to use oneof instead of repeated answer list

Leaving a proto list empty raises the potential ambiguity of an empty proto value, which can make it harder to reason about the model.

By using oneof, we can better control the structure of the message and make it clearer what fields are relevant to each question type.

adhiamboperes committed:
Refactor SurveyQuestion to have a language code

We can use this field to filter open-ended responses in analytics

adhiamboperes committed:
Add BUILD definitions for survey.proto

adhiamboperes committed:
Refactor SurveyUserAnswer to contain a oneof of the answer enums

adhiamboperes committed:
Refactored messages for readability and consistency.

adhiamboperes committed:
Remove unused import in oppia_logger.proto

adhiamboperes committed:
Rename enum field causing conflicts

STARTED_NOT_COMPLETED is already used in topic.

adhiamboperes committed:
Add more tests for topic active time

Test that the correct aggregate time is always saved.

adhiamboperes committed:
Merge branch 'nps-survey-models' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

adhiamboperes committed:
Refactor field names

Also fixed some spacing issues and updated comments.

adhiamboperes committed:
Merge branch 'nps-survey-models' of github.com:oppia/oppia-android into nps-survey-gating-logic

BenHenning reviewed (commented):
Thanks @adhiamboperes. A few things:
- If this is blocked on another PR to go in first, please update the PR description to include that information (e.g. "[Blocked: #4914]" at the end).
- If the PR description or PR code isn't complete yet, or if CI checks are failing prior to initial review, we should mark the PR as draft. (It's okay for check base branch check to fail, though, since this is a chained PR).
- Since this PR seems to be in draft (per the incomplete description & commented out code), please clarify what specific parts you'd like me to look at. I don't usually perform full reviews on draft PRs since they have a high chance of changing, so it's usually better to get the PR to code completion before a full review. However, I'm happy to look at & provide feedback on specific parts that you're unsure of (which can be especially useful if those are areas of code that are likely to change with feedback).

adhiamboperes committed:
Refactor dagger graph so that ExplorationProgressListener is not provided by the app component.

This does not need to be provided at application initialization.

adhiamboperes committed:
Refactor the implementation of ExplorationProgressListener

onExplorationSessionPaused receives the correct instances of topic ID and profile ID

adhiamboperes committed:
Refactor ExplorationProgressController to fire session start/end events

ExplorationProgressController will fire an ExplorationProgressListener callback whenever an exploration is started or exited.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into develop

adhiamboperes committed:
Remove commented out code

adhiamboperes commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationActiveTimeControllerTest.kt:
```diff
@@ -0,0 +1,219 @@
+    )
+  }
+
+  private fun createTopicLearningTimeDatabase(lastUpdatedTimeMs: Long): TopicLearningTimeDatabase {
```
I couldn't think of a way to enforce when the last record was updated, so I created this function. Could you suggest an alternative way that I can use to test this more robustly?

BenHenning replied:
As discussed in an other comment, this probably isn't ideal since it leaks an implementation detail of the controller (using PersistentCacheStore), and tests are most robust when they focus on verifying a public API's contract.

adhiamboperes commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationActiveTimeControllerTest.kt:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  @Test
+  fun testSaveSessionLength_previousAggregateIsStale_overwritesOldAggregateWithCurrentLength() {
```
This test always fails and I am unable to figure out why.
I have tried to debug using breakpoints, and the the code always hits the wrong if/else branch -- i.e the one that evaluates to timestamp not stale, and I cannot tell whether it is an issue with test set up or API. The API looks correct to me.

BenHenning replied:
As discussed & found during the meeting, it seems the failure is due to the PersistentCacheStore not actually writing the expected contents ahead of expecting the value to be available. I also suggest following a pattern more like AppStartupStateControllerTest wherein we only use the ExplorationActiveTimeController's API for setting & verifying state (since we can use two instances of the test suite's TestApplicationComponent to simulate existing state from a "previous app run").

adhiamboperes replied:
I was able to refactor all the tests to use the public apis by simulating the previous run suggested. Thanks.

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/ApplicationLifecycleListener.kt:
```diff
@@ -0,0 +1,10 @@
+package org.oppia.android.domain.oppialogger.analytics
+
+/** Listener for when the app goes to the foreground or the background. */
+interface ApplicationLifecycleListener {
```
One implementation of these listeners that I attempted was to implement it in `ApplicationLifecycleObserver`, similar to [ApplicationStartupListener](https://github.com/oppia/oppia-android/blob/1ba98ddb66590e7879ec01928ac6e59cd9da4408/domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/ApplicationLifecycleObserver.kt#L80). The only reason it does not work is that an implementation to start and stop a timer here will be triggered even when the user is not currently in the exploration screen.

I am not certain how best to enforce a check for current screen when the app is in foreground, to ensure the timer is only started when in ExplorationActivity.

I think the alternative would be to use the activity's lifecycle onPause/onResume. What do you think about this.

BenHenning replied:
As discussed in meeting today, I suggest refactoring the existing ApplicationLifecycleObserver such that we have a general handler for forwarding these lifecycle events, e.g.:

```kotlin
class ApplicationLifecycleObserver @Inject constructor(
  private val applicationLifecycleListeners: Set<ApplicationLifecycleListener>
): ApplicationStartupListener, LifecycleObserver {
  override fun onCreate() {
    ProcessLifecycleOwner.get().lifecycle.addObserver(this)
  }

  @OnLifecycleEvent(Lifecycle.Event.ON_START)
  fun onAppInForeground() {
    applicationLifecycleListeners.forEach(ApplicationLifecycleListener::onAppInForeground)
  }

  /** Occurs when application goes to background. */
  @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
  fun onAppInBackground() {
    applicationLifecycleListeners.forEach(ApplicationLifecycleListener::onAppInBackground)
  }
}
```

adhiamboperes replied:
I did this refactor and implemented the listeners in the `ExplorationProgressListenerImpl` as discussed, but I encontered these situations:
1. DRY: if the callbacks from both interfaces do the same thing, can we consolidate their implementations?
2. Not sure how the onAppInBackground() callback will receive the `profileId` and `topicId` needed when calling `setExplorationSessionPaused(profileId, topicId)` the callback itself takes no arguments, which is ideal since we fire it in the ApplicationLifecycleObserver like in your example above.

Here is what the impl looks like now:

```
@Singleton
class ExplorationProgressListenerImpl @Inject constructor(
  private val explorationActiveTimeController: ExplorationActiveTimeController
) : ExplorationProgressListener, ApplicationLifecycleListener {
  override fun onExplorationSessionStarted() {
    explorationActiveTimeController.setExplorationSessionStarted()
  }

  override fun onExplorationSessionPaused(profileId: ProfileId, topicId: String) {
    explorationActiveTimeController.setExplorationSessionPaused(profileId, topicId)
  }

  override fun onAppInForeground() {
    explorationActiveTimeController.setExplorationSessionStarted()
  }

  override fun onAppInBackground() {
    print("app is backgrounded")
  }
}
```
This way, with some logging, I can see that  `explorationActiveTimeController.setExplorationSessionStarted() `is always called, as soon as the app  comes into the foreground, regardless of the screen.

3. Where can I move this implementation so that the background/foreground behaviour is only reacted to within an exploration context?

BenHenning replied:
I think (2) and (3) relate to another comment I left in the listener impl--the correct way to do this is to probably use the signals of session activity & app activity independently to determine the current session state and use that for decision making. I think this also addresses (1) since I see the two listeners as providers of completely distinct information that we're using together to determine a particular domain state from which we can make decisions.

adhiamboperes replied:
I modified the implementation so that I receive signals from the two listeners independently then use them to decide the timer state.

adhiamboperes committed:
Refactor the save agg time api

Make the code more readable

adhiamboperes committed:
Refactor all TestApplicationComponents to have ExplorationProgressModule

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-update-test-components

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationActiveTimeControllerTest.kt:
```diff
@@ -0,0 +1,219 @@
+    // set aggregate learning time last updated timestamp to yesterday.
+    val lastUpdatedTimeMs = currentTime.minus(TimeUnit.DAYS.toMillis(1))
+
+    createCacheStore(lastUpdatedTimeMs)
```
```suggestion
    explorationActiveTimeController.setExplorationSessionStarted()
    testCoroutineDispatchers.advanceTimeBy(TimeUnit.DAYS.toMillis(1))
    explorationActiveTimeController.setExplorationSessionPaused(profileId, TEST_TOPIC_ID_0)
```

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -111,7 +111,8 @@ class ExplorationProgressController @Inject constructor(
   private val learnerAnalyticsLogger: LearnerAnalyticsLogger,
-  @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher
+  @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  private val explorationProgressListener: ExplorationProgressListener
```
```suggestion
  private val explorationProgressListeners: Set<explorationProgressListener>
```

adhiamboperes replied:
Got it, thanks!
An interesting thing I encountered was the need to annotate the type argument here with @JvmSuppressWildcards as the [documentation](https://dagger.dev/dev-guide/multibindings.html) warns.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressModule.kt:
```diff
@@ -0,0 +1,13 @@
+import dagger.Provides
+
+/** Module to provide dependencies corresponding to exploration progress. */
+@Module
+class ExplorationProgressModule {
+  @Provides
+  fun provideExplorationProgressListener(
+    explorationProgressListenerImpl: ExplorationProgressListenerImpl
+  ): ExplorationProgressListener = explorationProgressListenerImpl
+}
```
```suggestion
@Module
interface ExplorationProgressModule {
  @Binds
  @IntoSet
  fun provideExplorationProgressListener(
    impl: ExplorationProgressListenerImpl
  ): ExplorationProgressListener
}
```

adhiamboperes replied:
I updated my module in this way

BenHenning commented:
@adhiamboperes I followed up to the comments you left. Please let me know if there's anything else you'd like me to look at here.

adhiamboperes committed:
Fix merge conflicts while pulling develop locally

adhiamboperes committed:
Merge branch 'develop' into nps-survey-gating-logic

# Conflicts:
#	model/src/main/proto/survey.proto

adhiamboperes committed:
Refactor api access modifiers

Change internal functions to private

adhiamboperes committed:
Add tests for ExplorationActiveTimeController

adhiamboperes committed:
Fix missing KDOCs

adhiamboperes committed:
Refactor module for correct multibinding implementation

adhiamboperes committed:
Refactor ApplicationLifecycleObserver to handle lifecycle change callbacks.

Also provide the ApplicationLifecycleListener in the ApplicationLifecycleModule.

adhiamboperes committed:
Update from develop

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into nps-survey-update-test-components

# Conflicts:
#	model/src/main/proto/survey.proto

adhiamboperes committed:
Merge branch 'nps-survey-update-test-components' into nps-survey-gating-logic

adhiamboperes committed:
Fix import ordering in affected test files

adhiamboperes committed:
Add KDOC for SUCCESS in TopicLearningTimeActionStatus

adhiamboperes committed:
Fix failing DateTimeUtilTest

adhiamboperes committed:
Add test file exemptions for listeners and modules.

adhiamboperes committed:
Add more enum fields to represent time of day

create more granularity by breaking up the hours to early morning, morning, evening and night. The aim of this change is to aid in gating that relies on time of day check.

adhiamboperes committed:
Refactor time of day gating logic to not use calendar directly

Make use of MachineLocale time of day function.

adhiamboperes committed:
Add test file for SurveyGatingController

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListenerImpl.kt:
```diff
@@ -0,0 +1,27 @@
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Implementation of [ExplorationProgressListener] */
```
Calling listener implementations generic "impl"s poses a problem of vagueness. Listeners are meant to exist more than once, so each implementation ought to be specific to whichever domain purpose it's fulfilling (versus utilities which generally only have one implementation). I suggest renaming this class to be more specific as to what it's responsible for rather than just naming it an implementation of the listener class (especially since it's an impl of multiple listeners, so there's clearly more specificity possible here).

adhiamboperes replied:
Since the purpose here is to start and stop the session timer, I have renamed it to `ExplorationSessionTimerController`

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListenerImpl.kt:
```diff
@@ -0,0 +1,27 @@
+    explorationActiveTimeController.setExplorationSessionStarted()
+  }
+
+  override fun onExplorationSessionPaused(profileId: ProfileId, topicId: String) {
+    explorationActiveTimeController.setExplorationSessionPaused(profileId, topicId)
+  }
```
Is this 'paused' or 'ended'?

adhiamboperes replied:
Ended.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListenerImpl.kt:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  override fun onAppInForeground() {
+    explorationActiveTimeController.setExplorationSessionStarted()
```
Isn't this only true if there's an ongoing session? I'd also expect onAppInBackground to play a role in 'pausing' logic.

I think either the listener needs to determine session state based on its various signals & pass this along to the controller, or it needs to forward its received values to the controller and let it perform the determination.

Also a separate thought: you could alternatively do away with this class entirely and just have the controller implement the listeners & bind to the corresponding listener sets.

adhiamboperes replied:
I have added a check for an exploration being in progress.

I have opted to keep this class, the listeners determine the session state and passes it along to the controller.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -570,6 +573,7 @@ class ExplorationProgressController @Inject constructor(
         recomputeCurrentStateAndNotifyAsync()
       }.launchIn(CoroutineScope(backgroundCoroutineDispatcher))
       explorationProgress.advancePlayStageTo(LOADING_EXPLORATION)
+      explorationProgressListeners.forEach(ExplorationProgressListener::onExplorationSessionStarted)
```
For this & the below callback: does exploration ID never need to be passed?

adhiamboperes replied:
At the moment, I am not using the exploration ID for any logic, which is why I am not passing it.
The granularity currently required is at topic id level.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -273,21 +275,56 @@ class ExplorationActivityPresenter @Inject constructor(
-              backPressActivitySelector()
-              (activity as ExplorationActivity).finish()
+              if (isCompletion) {
+                shouldShowSurveyDialog(profileId, topicId)
```
Perhaps instead: maybeShowSurveyDialog? The current name reads as a function that will return whether to show it as opposed to actually showing it.

adhiamboperes replied:
I'll keep this in mind for when I implement the dialog, although this will be in theb next PR in the chain. I have already removed this placeholder code.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -273,21 +275,56 @@ class ExplorationActivityPresenter @Inject constructor(
       )
   }
 
+  private fun shouldShowSurveyDialog(profileId: ProfileId, topicId: String) {
+    surveyGatingController.shouldShowSurvey(profileId, topicId).toLiveData()
+      .observe(
+        activity,
+        {
+          when (it) {
+            is AsyncResult.Pending -> {
+              oppiaLogger.d("ExplorationActivity", "Checking survey gating")
+            }
+            is AsyncResult.Failure -> {
+              oppiaLogger.e(
+                "ExplorationActivity",
+                "Failed to retrieve gating decision",
+                it.error
+              )
+            }
+            is AsyncResult.Success -> {
+              if (it.value) {
+                createSurveyDialog().show()
+              } else {
+              }
+            }
+          }
+        }
+      )
+  }
+
+  private fun createSurveyDialog(): AlertDialog {
+    val builder = AlertDialog.Builder(activity)
+    return builder.setTitle("NPS Survey")
+      .setMessage("Survey Intro message")
+      .create()
+  }
```
I assume this is temporary as it'll be filled in properly in a later PR? Reminder to remove this code in that case.

adhiamboperes replied:
Yes, I have removed the code.

BenHenning commented on model/src/main/proto/profile.proto:
```diff
@@ -80,6 +80,9 @@ message Profile {
   // Determines if the in-exploration continue button animation has been seen by the user or not.
   bool is_continue_button_animation_seen = 15;
+
+  // Represents the last time the nps survey was shown on this profile.
+  int64 survey_last_shown_timestamp = 17;
```
Please update comment & field name to indicate the time unit being used.

adhiamboperes replied:
I updated this.

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -671,3 +671,23 @@ message SubtopicRecord {
 }
+
+// Top-level proto used to store the aggregate learning duration timestamps in a topic,
+// per-profile basis.
```
```suggestion
// on per-profile basis.
```

adhiamboperes replied:
Noted

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -671,3 +671,23 @@ message SubtopicRecord {
+  // The ID of the topic for which this learning time has been recorded.
+  string topic_id = 1;
+
+  // The total time spent by a learner in this topic.
```
```suggestion
  // The total time spent by a learner in this topic, in milliseconds.
```

adhiamboperes replied:
Updated.

BenHenning commented on model/src/main/proto/topic.proto:
```diff
@@ -671,3 +671,23 @@ message SubtopicRecord {
+  int64 topic_learning_time_ms = 2;
+
+  // The last time the topic_aggregate_learning_time_ms was updated, coinciding with the last time
+  // the learner played an exploration in this topic.
```
I suggest indicating that this is a Unix epoch timestamp in milliseconds (as opposed to a duration or any other type of time value that could be stored here).

adhiamboperes replied:
I updated this comment

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -646,6 +646,8 @@ exempted_file_path: "domain/src/main/java/org/oppia/android/domain/classify/rule
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/classify/rules/textinput/TextInputRuleModule.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgress.kt"
+exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListener.kt"
+exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressModule.kt"
```
This actually could be tested--I suggest adding a basic test suite to demonstrate that the binding is set up correctly & as expected by the module (we have a few other module tests in the codebase that you could look at for reference).

adhiamboperes replied:
Noted, thanks.

BenHenning commented on utility/src/main/java/org/oppia/android/util/locale/OppiaLocale.kt:
```diff
@@ -155,6 +155,9 @@ sealed class OppiaLocale {
+      EARLY_MORNING,
+
       /** Corresponds to the user's morning time. */
       MORNING,
```
Perhaps this should be 'MID_MORNING' or 'LATE_MORNING' to contrast with 'EARLY_MORNING'?

adhiamboperes replied:
I went with MID_MORNING, given the range of hours under consideration.

BenHenning commented on utility/src/main/java/org/oppia/android/util/locale/MachineLocaleImpl.kt:
```diff
@@ -50,9 +50,11 @@ class MachineLocaleImpl @Inject constructor(
       in 12..16 -> TimeOfDay.AFTERNOON
-      in 0..3, in 17..23 -> TimeOfDay.EVENING
+      in 17..22 -> TimeOfDay.EVENING
+      in 0..3, 23 -> TimeOfDay.NIGHT
```
Perhaps 'LATE_NIGHT'?

adhiamboperes replied:
Yes, this is definitely more appropriate.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt:
```diff
@@ -0,0 +1,89 @@
+  private val machineLocale: OppiaLocale.MachineLocale
+) {
+
+  private val surveyGracePeriodMillis = TimeUnit.DAYS.toMillis(30)
```
Is this something we should maybe implement as a platform parameter? It could be useful to be able to remotely tweak it in the future (and platform parameters make it easier to swap out this value while testing).

adhiamboperes replied:
I did find out from testing that platform parameters would be preferrable here.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes!

I actually realized a bit late that this PR is still in draft (so I was reviewing it as though it were fully in review). I stopped short of reviewing the controllers in depth, and the tests at all. I took a look at everything else, though, and left some thoughts. PTAL.

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into nps-survey-gating-logic

adhiamboperes committed:
Refactor: remove placeholder alert dialog for survey.

Was using this to test that gating actually works.

adhiamboperes committed:
Update naming and comments of time related proto fields

To make it clear which time unit is being used.

adhiamboperes committed:
Update TimeOfDay enum field naming

adhiamboperes committed:
Refactor disk IO operations when fetching gating criteria to Async.

Previously the operations were happening synchronously.

adhiamboperes committed:
Refactor gating API.

Use combineWith operations to transform the gating criteria and return a gating result as a boolean

adhiamboperes committed:
Add tests for surveyLastShownTimestampMs in ProfileManagementController

Test that the save and retrieve APIs work as expected.

adhiamboperes committed:
Refactor method names

Update method names and comments to be more semantic.

adhiamboperes committed:
Add bazel BUILD definitions for new survey related files.

Added //domain to prod_module because at the moment, files in existing packages without own build files get added to top-level domain module, and we depend on this `domain` module when whe need to access those files.

adhiamboperes committed:
Assorted refactors

Add bazel BUILD file for SurveyGatingController
Remove unnecessary test file exemption
Rename test constants to correspond with prod names
Remove extra test deps

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Assorted refactors

Add bazel BUILD file for SurveyGatingController
Remove unnecessary test file exemption
Rename test constants to correspond with prod names
Fix lint warnings

adhiamboperes committed:
Refactor and complete session timer logic that is based on application lifecycle changes.

Here we are only using topicId and not explorationId because of the level of granularity required currently is at topic level.

adhiamboperes committed:
Add survey gating criteria constants as platform parameters.

This is so that they can be adjusted remotely if needed

adhiamboperes committed:
Add gating test outlines: wip

Missing complete scenario setup and assertions

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Fix bazel test failure on class ExplorationActiveTimeControllerTest

Added missing module dependency

adhiamboperes committed:
Add test for ExplorationProgressModule

Verify that the listener set is bound correctly.

adhiamboperes committed:
Fix formatting error

adhiamboperes committed:
Cleanup unused fields.

adhiamboperes committed:
Add test for providing ApplicationLifecycleListener set

adhiamboperes committed:
Add test file exemption for ExplorationSessionTimerController

adhiamboperes committed:
Fix failing InitializeDefaultLocaleRule tests

adhiamboperes committed:
Add test definitions for SurveyGatingControllerTest

BenHenning commented:
Apologies--will need to look at this tomorrow.

Also @adhiamboperes it looks like there are CI failures--it might be good to address these first.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Add extensive tests for gating controller

Also reorganized tests for readability

adhiamboperes committed:
Add survey gating platform params to test PlatformParam module

To easily provide these platform parameters in a test.

adhiamboperes committed:
Fix gating logic error found in tests

Correctly compute whether the Survey Grace Period has Ended

adhiamboperes committed:
Fix bazel test errors

remove unused modules, add required modules and definitions

adhiamboperes committed:
Add missing imports in TestApplicationComponent

These we causing tests failures on bazel for the instrumentation package.

adhiamboperes commented:
> Also @adhiamboperes it looks like there are CI failures--it might be good to address these first.

Hi @BenHenning, I have resolved the CI failures.

adhiamboperes committed:
Merge branch 'develop' into nps-survey-gating-logic

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt:
```diff
@@ -0,0 +1,182 @@
+   * @param profileId the ID corresponding to the profile for which progress needs to be stored
+   * @param topicId the ID corresponding to the topic for which duration needs to be stored
+   * @param sessionDuration the tracked exploration duration between start and pause
+   * @return a [DataProvider] that indicates the success/failure of this record operation
```
Punctuation?

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -530,7 +531,9 @@ class ExplorationProgressController @Inject constructor(
       }
-    } catch (e: Exception) { AsyncResult.Failure(e) }
+    } catch (e: Exception) {
+      AsyncResult.Failure(e)
```
Revert to one line

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListener.kt:
```diff
@@ -0,0 +1,12 @@
+
+import org.oppia.android.app.model.ProfileId
+
+/** Listener for when an exploration is started or paused. */
```
`/** Listener for when an exploration is started or ended. */`

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/ApplicationLifecycleModule.kt:
```diff
@@ -18,4 +19,10 @@ class ApplicationLifecycleModule {
+  @Provides
+  @IntoSet
+  fun provideApplicationLifecycleListeners(
+    impl: ExplorationSessionTimerController
```
Rename argument

adhiamboperes committed:
Merge branch 'develop' into nps-survey-gating-logic

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterAlphaKenyaModule.kt
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt
#	testing/src/main/java/org/oppia/android/testing/platformparameter/TestPlatformParameterModule.kt
#	utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt

adhiamboperes committed:
Refactor tests for clarity

adhiamboperes committed:
Add test for edge case

adhiamboperes committed:
Fix merge conflicts from develop

adhiamboperes commented on instrumentation/src/java/org/oppia/android/instrumentation/application/TestApplicationComponent.kt:
```diff
@@ -100,7 +101,7 @@ import javax.inject.Singleton
     PerformanceMetricsAssessorModule::class, PerformanceMetricsConfigurationsModule::class,
     TestingBuildFlavorModule::class, EventLoggingConfigurationModule::class,
-    CpuPerformanceSnapshotterModule::class
+    ExplorationProgressModule::class, CpuPerformanceSnapshotterModule::class,
```
Remove comma

adhiamboperes commented on model/src/main/proto/profile.proto:
```diff
@@ -80,6 +80,9 @@ message Profile {
   // Determines if the in-exploration continue button animation has been seen by the user or not.
   bool is_continue_button_animation_seen = 15;
+
+  // Represents the timestamp in milliseconds when the nps survey was shown on this profile.
```
```suggestion
  // Represents the epoch timestamp in milliseconds when the nps survey was previously shown on this profile.
```

adhiamboperes commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -646,8 +646,10 @@ exempted_file_path: "domain/src/main/java/org/oppia/android/domain/classify/rule
+exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListener.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationRetriever.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationRetrieverImpl.kt"
+exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationSessionTimerController.kt"
```
Remove excemption and actually add tests

adhiamboperes commented on utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt:
```diff
@@ -231,3 +232,42 @@ annotation class EnableInteractionConfigChangeStateRetention
+ * Default value of the minimum learning time in a topic, in minutes, that indicates a learner has
+ * had sufficient interaction with the app to be able to provide informed feedback about their
+ * experience with the app.
+ * */
```
Fix KDoc formatting

adhiamboperes commented on utility/src/test/java/org/oppia/android/util/locale/MachineLocaleImplTest.kt:
```diff
@@ -299,12 +317,18 @@ class MachineLocaleImplTest {
   private companion object {
     // Date & time: Wed Apr 24 2019 08:22:03 GMT.
-    private const val MORNING_UTC_TIMESTAMP_MILLIS = 1556094123000
+    private const val EARLY_MORNING_UTC_TIMESTAMP_MILLIS = 1556094123000
```
Verify tests in this class for exhaustiveness per new classifications

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationSessionTimerController.kt:
```diff
@@ -0,0 +1,58 @@
+
+  override fun onAppInBackground() {
+    this.isAppInForeground = false
+    if (profileId != null && topicId != null) {
```
Test these changes

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -570,6 +573,12 @@ class ExplorationProgressController @Inject constructor(
       }.launchIn(CoroutineScope(backgroundCoroutineDispatcher))
       explorationProgress.advancePlayStageTo(LOADING_EXPLORATION)
+      explorationProgressListeners.forEach {
+        it.onExplorationStarted(
```
Test that these listeners are called.

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt:
```diff
@@ -0,0 +1,182 @@
+   * This method is called when the [ExplorationProgressController.finishExplorationImpl] finishes
+   * executing, or the app goes to the background.
+   */
+  fun setExplorationSessionStopped(profileId: ProfileId, topicId: String) {
```
What happens if this is called before a session was started?

adhiamboperes reviewed (commented):
Adding some self-review comments.

adhiamboperes reviewed (commented):
Adding some self-review comments.

adhiamboperes committed:
Add tests for ExplorationSessionTimerController and remove it from exemptions

adhiamboperes committed:
Fix session timer control logic

Verify logic via tests

adhiamboperes committed:
Fix some nits from self-review

- Argument names
- KDocs formatting
- KDoc wording
- General formatting
- Import ordering

adhiamboperes committed:
Remove check that caused a failure in an existing test

adhiamboperes committed:
Fix bazel dependency out-of-order-load

adhiamboperes committed:
Refactor: reorder deps in BUILD.bazel

adhiamboperes committed:
Merge branch 'develop' into nps-survey-gating-logic

adhiamboperes committed:
Reformat BUILD file

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt:
```diff
@@ -135,11 +133,11 @@ class ExplorationActiveTimeController @Inject constructor(
+      .transformAsync(RETRIEVE_AGGREGATE_LEARNING_TIME_PROVIDER_ID) { learningTimeDb ->
+        return@transformAsync AsyncResult.Success(
+          learningTimeDb.aggregateTopicLearningTimeMap[topicId]
+            ?: TopicLearningTime.getDefaultInstance()
```
This can just be transform() if you're not calling a suspend function or returning a non-success AsyncResult.

adhiamboperes replied:
Changed this to transform as it has no suspend context

kkmurerwa replied:
This looks good now.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt:
```diff
@@ -80,7 +78,7 @@ class ExplorationActiveTimeController @Inject constructor(
    * @param topicId the ID corresponding to the topic for which duration needs to be stored
    * @param sessionDuration the tracked exploration duration between start and pause
-   * @return a [DataProvider] that indicates the success/failure of this record operation
+   * @return a [DataProvider] that indicates the success/failure of this record operation.
```
Nit: no punctuation at the end of an at-clause's fragment (we only add punctuation if multiple sentences are needed).

adhiamboperes replied:
Fixed

kkmurerwa replied:
This looks good to me.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListener.kt:
```diff
@@ -2,11 +2,11 @@ package org.oppia.android.domain.exploration
 import org.oppia.android.app.model.ProfileId
 
-/** Listener for when an exploration is started or paused. */
+/** Listener for when an exploration is started or ended. */
```
```suggestion
/** Listener for when an exploration has started or ended. */
```

Nit on verb use.

adhiamboperes replied:
Corrected

kkmurerwa replied:
This is good now.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/profile/ProfileManagementControllerTest.kt:
```diff
@@ -1108,6 +1112,75 @@ class ProfileManagementControllerTest {
+      )
+    )
+
+    assertThat(lastShownTimeMs).isEqualTo(oppiaClock.getCurrentTimeMs())
```
I suggest setting the time explicitly at the start of the test to make it clear that it's the expected current time at this point (since it's not obvious at a glance that getCurrentTimeMs() shouldn't return DEFAULT_SURVEY_LAST_SHOWN_TIMESTAMP_MILLIS which, if it did, could result in one or both of these tests falsely passing in cases when the production code regresses).

adhiamboperes replied:
Thanks for this suggestion, it does make a lot of sense.

kkmurerwa replied:
I've reviewed the updated `testFetchSurveyLastShownTime_updateLastShownTimeFunctionCalled_returnsCurrentTime` test and it should work well now without the potential risk of false positives.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/oppialogger/analytics/ApplicationLifecycleModuleTest.kt:
```diff
@@ -66,6 +68,10 @@ class ApplicationLifecycleModuleTest {
+  @Test
+  fun testInjectApplicationLifecycleListenerSet_includesExplorationSessionTimerController() {
+    assertThat(lifecycleListeners.any { it is ExplorationSessionTimerController }).isTrue()
+  }
```
Nit: newline after the brace for member spacing in the class.

adhiamboperes replied:
Fixed

kkmurerwa replied:
I am unable to confirm whether the suggested spacing was added.

adhiamboperes replied:
There was previously no newline at ln78. I added it.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationSessionTimerControllerTest.kt:
```diff
@@ -0,0 +1,332 @@
+    )
+    testCoroutineDispatchers.runCurrent()
+
+    testCoroutineDispatchers.advanceTimeBy(SESSION_LENGTH)
```
Here & elsewhere: you can omit the runCurrent before thi scall since advanceTimeBy implies runCurrent (since it runs all tasks up to the scheduled delay).

adhiamboperes replied:
I ommitted all the occurance of runCurrent except for one instance where it is actually needed. In the test `testSessionTimer_explorationStarted_appInBackground_thenInForeground_resumesSessionTimer`, I need all previous operations to run before advancing time. I am using advanceTimeBy to get the clock to  a specific time, as opposed to changing the FakeTimeMode to MODE_FIXED_FAKE_TIME and manually setting time.

kkmurerwa replied:
All `runCurrent()` instances with an `advanceTimeBy()` method just below them seem to have been updated. This is good to go.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt:
```diff
@@ -0,0 +1,102 @@
+    showNextCalendar.timeInMillis = showNextTimestamp
+
+    return currentCalendar.after(showNextCalendar) ||
+      showNextCalendar == currentCalendar
```
Is a ``Calendar`` really needed for this logic? I'm wondering if it's just checking/managing the Unix timestamp internally. This might be able to be simplified if so since then we can just avoid using ``Calendar`` and work with the timestamps directly.

adhiamboperes replied:
I was able to compare the timestamps directly without using calendar. I refactored the method.

kkmurerwa replied:
This looks resolved.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt:
```diff
@@ -0,0 +1,102 @@
+    return profileManagementController.retrieveSurveyLastShownTimestamp(profileId)
+      .transformAsync(GET_SURVEY_LAST_SHOWN_TIMESTAMP_PROVIDER_ID) { lastShownTimestampMs ->
+        AsyncResult.Success(lastShownTimestampMs)
+      }
```
Can this not just be ``retrieveSurveyLastShownTimestamp()``? The transformation doesn't seem to be changing anything.

adhiamboperes replied:
I made this change

kkmurerwa replied:
This is good now.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt:
```diff
@@ -0,0 +1,102 @@
+    ).transformAsync(GET_TOPIC_LEARNING_TIME_PROVIDER_ID) { topicLearningTime ->
+      val aggregateLearningTimeMs = topicLearningTime.topicLearningTimeMs
+      AsyncResult.Success(aggregateLearningTimeMs)
+    }
```
```suggestion
    return explorationActiveTimeController.retrieveAggregateTopicLearningTimeDataProvider(
      profileId, topicId
    ).transform(
      GET_TOPIC_LEARNING_TIME_PROVIDER_ID, TopicLearningTime::getTopicLearningTimeMs
    )
```

Can be simplified a bit (and probably shortened if some of the variable names involved were shortened).

adhiamboperes replied:
I did a refactor per the suggestion

kkmurerwa replied:
The changes made look good.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationSessionTimerController.kt:
```diff
@@ -0,0 +1,59 @@
+class ExplorationSessionTimerController @Inject constructor(
+  private val explorationActiveTimeController: ExplorationActiveTimeController
+) : ExplorationProgressListener, ApplicationLifecycleListener {
+  private var isAppInForeground: Boolean = false
+  private var explorationStarted: Boolean = false
+  private var profileId: ProfileId? = null
+  private var topicId: String? = null
```
It seems like these variables will be accessed on different threads. These may need to be transferred over to an internal data class combined with either a command pattern (like we use in ``ExplorationProgressController`` via an actor) or to ``InMemoryBlockingCache`` to ensure synchronous access to the data).

adhiamboperes replied:
I chose to use the command pattern approach since I could not find sufficient references to InMemoryBlockingCache to learn how to use it, due to limited use in the codebase.

This refactor, though it ended up being non-trivial, allowed me to have better control over the timer, since the timer now has its own session that I can start, pause, resume and stop.

kkmurerwa replied:
The logic for this seems to have been moved to the `ExplorationActiveTimeController` class which is consistent with the other changes. The approach you chose also LGTM.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationSessionTimerController.kt:
```diff
@@ -0,0 +1,59 @@
+ */
+@Singleton
+class ExplorationSessionTimerController @Inject constructor(
+  private val explorationActiveTimeController: ExplorationActiveTimeController
+) : ExplorationProgressListener, ApplicationLifecycleListener {
```
Why not have the ``ExplorationActiveTimeController`` itself bind as an instance of these listeners? It seems like the extra wrapper doesn't add much benefit (unless I'm missing something).

adhiamboperes replied:
I refactored this and removed ExplorationSessionTimerController and its test file

kkmurerwa replied:
This looks done.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressListener.kt:
```diff
@@ -0,0 +1,12 @@
+import org.oppia.android.app.model.ProfileId
+
+/** Listener for when an exploration is started or ended. */
+interface ExplorationProgressListener {
```
It's worthwhile to document whether this listener may be accessed on multiple threads (so that implementers can take precautions, if needed).

adhiamboperes replied:
I updated this Kdoc

kkmurerwa replied:
This looks better now. It can be marked as resolved.

kkmurerwa replied:
This has now been addressed.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -570,6 +571,12 @@ class ExplorationProgressController @Inject constructor(
         recomputeCurrentStateAndNotifyAsync()
       }.launchIn(CoroutineScope(backgroundCoroutineDispatcher))
       explorationProgress.advancePlayStageTo(LOADING_EXPLORATION)
+      explorationProgressListeners.forEach {
```
I suggest adding tests to ``ExplorationProgressControllerTest`` to verify both the session start & end cases to make sure that bound listeners are called when expected and passed the correct data.

adhiamboperes replied:
I added one test for this

kkmurerwa replied:
I've reviewed the added test and it covers the required cases. This should be good to go.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt:
```diff
@@ -0,0 +1,182 @@
+  private fun retrieveCacheStore(
+    profileId: ProfileId
+  ): PersistentCacheStore<TopicLearningTimeDatabase> {
+    val cacheStore = if (profileId in cacheStoreMap) {
+      cacheStoreMap[profileId]!!
+    } else {
+      val cacheStore =
+        cacheStoreFactory.createPerProfile(
+          CACHE_NAME,
+          TopicLearningTimeDatabase.getDefaultInstance(),
+          profileId
+        )
+      cacheStoreMap[profileId] = cacheStore
+      cacheStore
+    }
+
+    cacheStore.primeInMemoryAndDiskCacheAsync(
+      updateMode = PersistentCacheStore.UpdateMode.UPDATE_IF_NEW_CACHE,
+      publishMode = PersistentCacheStore.PublishMode.PUBLISH_TO_IN_MEMORY_CACHE
+    ).invokeOnCompletion {
+      if (it != null) {
+        oppiaLogger.e(
+          "ExplorationActiveTimeController",
+          "Failed to prime cache ahead of data retrieval.",
+          it
+        )
+      }
+    }
+    return cacheStore
```
```suggestion
    return cacheStoreMap.getOrPut(profileId) {
      cacheStoreFactory.createPerProfile(
        CACHE_NAME,
        TopicLearningTimeDatabase.getDefaultInstance(),
        profileId
      ).also { cacheStore ->
        cacheStore.primeInMemoryAndDiskCacheAsync(
          updateMode = PersistentCacheStore.UpdateMode.UPDATE_IF_NEW_CACHE,
          publishMode = PersistentCacheStore.PublishMode.PUBLISH_TO_IN_MEMORY_CACHE
        ).invokeOnCompletion {
          if (it != null) {
            oppiaLogger.e(
              "ExplorationActiveTimeController",
              "Failed to prime cache ahead of data retrieval.",
              it
            )
          }
        }
      }
    }
```

Shorter, more idiomatic, and avoids the potential code smell of a forced null check.

adhiamboperes replied:
I accepted this suggestion as is, thanks!

kkmurerwa replied:
This can be marked as resolved.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt:
```diff
@@ -0,0 +1,182 @@
+    val deferred = retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) { topicLearningTimeDatabase ->
+
+      val previousAggregateLearningTime =
+        topicLearningTimeDatabase.aggregateTopicLearningTimeMap[topicId]
+
+      val topicLearningTimeBuilder = if (previousAggregateLearningTime != null) {
+        previousAggregateLearningTime.toBuilder()
+      } else {
+        TopicLearningTime.newBuilder()
+          .setTopicId(topicId)
+          .setLastUpdatedTimeMs(oppiaClock.getCurrentTimeMs())
+      }
+
+      if (isLastUpdatedTimestampStale(topicLearningTimeBuilder.lastUpdatedTimeMs)
+      ) {
+        topicLearningTimeBuilder.topicLearningTimeMs = sessionDuration
+      } else {
+        topicLearningTimeBuilder.topicLearningTimeMs += sessionDuration
+      }
+      topicLearningTimeBuilder.lastUpdatedTimeMs = oppiaClock.getCurrentTimeMs()
+
+      val topicLearningTime = topicLearningTimeBuilder.build()
+
+      val topicLearningTimeDatabaseBuilder = topicLearningTimeDatabase.toBuilder()
+        .putAggregateTopicLearningTime(topicId, topicLearningTime)
+      Pair(topicLearningTimeDatabaseBuilder.build(), TopicLearningTimeActionStatus.SUCCESS)
```
```suggestion
      topicLearningTimeDatabase.toBuilder().apply {
        val topicLearningTime = aggregateTopicLearningTimeMap.getOrDefault(topicId) {
          TopicLearningTime.newBuilder().setTopicId(topicId).build()
        }.toBuilder().apply {
          topicLearningTimeMs = if (isLastUpdatedTimestampStale(lastUpdatedTimeMs)) {
            sessionDuration
          } else topicLearningTimeMs + sessionDuration
          lastUpdatedTimeMs = oppiaClock.getCurrentTimeMs()
        }.build()
        putAggregateTopicLearningTime(topicId, topicLearningTime)
      }.build() to TopicLearningTimeActionStatus.SUCCESS
```

Bunch of small idiomatic simplifications.

adhiamboperes replied:
I accepted this suggestion with some minor changes.

kkmurerwa replied:
The changes LGTM.

BenHenning reviewed (approved):
Thanks @adhiamboperes! I Took a full pass on the PR. While I am leaving a fair number of suggestions, many of these are actually superficial architectural suggestions or nits. I do suggest addressing all of them, and perhaps having another team member take a pass on the changes.

That being said, approving the PR since it largely looks good to me and I don't want to block this getting merged while I'm not performing code reviews over the coming weeks.

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into nps-survey-gating-logic

# Conflicts:
#	model/src/main/proto/profile.proto

adhiamboperes committed:
Refactor cachestore read and write methods to be shorter

Fixes:
1. punctuation in kdocs: https://github.com/oppia/oppia-android/pull/4916#discussion_r1213870876

2. Change from transformAsync to transform since this is not a suspend function: https://github.com/oppia/oppia-android/pull/4916#discussion_r1213870629

3. Shorten read operation: https://github.com/oppia/oppia-android/pull/4916#discussion_r1214031917

4. Shorten write operation:https://github.com/oppia/oppia-android/pull/4916#discussion_r1214088237

adhiamboperes committed:
Refactor unused transformation to direct call

Fixes: https://github.com/oppia/oppia-android/pull/4916#discussion_r1214030027

adhiamboperes committed:
Refactor retrieve retrieveAggregateLearningTime transformation to synchronous

Fixes: https://github.com/oppia/oppia-android/pull/4916#discussion_r1214031917

adhiamboperes committed:
Fix missing default value in recordAggregateTopicLearningTime()

adhiamboperes committed:
Refactor grace period computation to compare timestamps directly without  using calendar

Fixes https://github.com/oppia/oppia-android/pull/4916#discussion_r1214029219

adhiamboperes committed:
Fix nit on verb use in ExplorationProgressListener

Fixes: https://github.com/oppia/oppia-android/pull/4916#discussion_r1213871142

adhiamboperes committed:
Add explicit time declaration before running tests

Setting the time explicitly at the start of the test makes it clear that it's the expected current time at this point (since it's not obvious at a glance that getCurrentTimeMs() shouldn't return DEFAULT_SURVEY_LAST_SHOWN_TIMESTAMP_MILLIS which, if it did, could result in one or both of these tests falsely passing in cases when the production code regresses).

Fixes https://github.com/oppia/oppia-android/pull/4916#discussion_r1213879231

adhiamboperes committed:
Add caution for accessing ExplorationProgressListener on multiple threads

Fixes https://github.com/oppia/oppia-android/pull/4916#discussion_r1214058001

adhiamboperes committed:
Refactor ExplorationActiveTimeController to bind as an instance of ApplicationLifecycleListener and ExplorationProgressListener

These do not need to be in a separate class/wrapper

Fixes:
1. https://github.com/oppia/oppia-android/pull/4916#discussion_r1214057560
2. Nit on newline after brace https://github.com/oppia/oppia-android/pull/4916#discussion_r1213879484

adhiamboperes committed:
Remove the now unused SessionTimerController and test file

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Fix KDoc formatting

adhiamboperes committed:
Refactor the begin session timer logic to a command pattern

This is to ensure synchronous access to the data (See https://github.com/oppia/oppia-android/pull/4916#discussion_r1214056747).

The associated tests have also been modified, and runCurrent() removed except where appropriate, per https://github.com/oppia/oppia-android/pull/4916#discussion_r1213879871

adhiamboperes committed:
Merge branch 'develop' into nps-survey-gating-logic

adhiamboperes committed:
Fix associations broken by refactor

adhiamboperes commented:
Hi @kkmurerwa, could you please take a look at the changes I have made on this PR(commits since last review).

I have attempted to address each comment on it's own commit, and the commit message  body should give you more information about what has changed.

adhiamboperes committed:
Fix gating controller tests broken by refactor

adhiamboperes committed:
Fix timer controller tests that run in 2 app instances

There is an issue with timing when running operations in a simulated PreviousAppInstance. I think that some scheduled async tasks never reach a reasonable idle state, so that it becomes impossible to access their outcome. I have used runCurrent here to force an idle state for those scenarios. This is also the approach used in https://github.com/oppia/oppia-android/blob/develop/domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt

adhiamboperes committed:
Fix failing test

- Correct the test setup
- Small naming fixes

adhiamboperes committed:
Add a test to verify that explorationProgressListeners are fired

Test that a session timer is initialized and finished by checking that the aggregate time was updated.

Also, a small refactor on calling the onExplorationEnded listener

kkmurerwa commented:
Hi @adhiamboperes. I will review it tomorrow over the course of the day.

adhiamboperes committed:
Fix failing tests

adhiamboperes committed:
Add missing file to vcs

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-gating-logic

adhiamboperes committed:
Create test file exemption for TimerSessionState

adhiamboperes committed:
Reformat translation/BUILD.bazel

adhiamboperes committed:
Reformat TimerSessionState KDoc

kkmurerwa commented:
Hi @adhiamboperes. I've looked over the comments to make sure all suggestions have been implemented and everything seems good. Once the failing `testGetCurrentState_playInvalidExploration_returnsFailure_logsException` test on the `ExplorationProgressControllerTest` is fixed, this PR should be good to go.

adhiamboperes committed:
Remove unnecessary condition before beginning session timer

adhiamboperes committed:
Merge branch 'develop' into nps-survey-gating-logic

adhiamboperes commented:
> Hi @adhiamboperes. I've looked over the comments to make sure all suggestions have been implemented and everything seems good. Once the failing `testGetCurrentState_playInvalidExploration_returnsFailure_logsException` test on the `ExplorationProgressControllerTest` is fixed, this PR should be good to go.

Thanks @kkmurerwa! I have fixed the failing test and resolved the remaining threads.

adhiamboperes commented:
@seanlip, could you please code-owner approve and merge this PR for me? Thanks.

seanlip commented:
Hm, @BenHenning seems to have approved it already, so no further approval is needed from me AFAICT. I'll just merge it!
